### PR TITLE
Initial Scala 3 support

### PR DESCRIPTION
**Pull request by pomadchin**
Connected https://github.com/typelevel/cats-tagless/issues/170

**TODO (an approximate list)**:

Supported instances derivation body functions:

- **FunctorK**: https://github.com/typelevel/cats-tagless/pull/442
  - [x] `A => F[_]`
  - [ ] `A => Kleisli / WriterT / WriterT / etc`
    - is there a generic solution?
  - [ ] `Kleisli / WriterT / WriterT / etc` => `A`
  - [x] `AlgWithNonEffectMethod`
  - [ ] `AlgWithTypeMember`
  - [x] `AlgWithTypeBound`
  - [x] `[X] => F[X] => F[_]` generic functions support
    - [x] `AlgWithTParamInMethod` 
  - [ ] `AlgWithAbstractTypeClass`
  - [x] `AlgWithCurryMethod`
  - [ ] Algebras with extra Type Parameters 
    - [ ] `AlgWithExtraTP`
    - [x] `AlgWithExtraTP2` 
  - [x] Varargs
    - [x]  `AlgWithVarArgsParameter`
  - [ ] `BuilderAlgebra`
- **SemigroupalK** https://github.com/typelevel/cats-tagless/pull/513, https://github.com/typelevel/cats-tagless/pull/516
  - [x] `A => F[_]`
  - [ ] `A => Kleisli / WriterT / WriterT / etc`
    - is there a generic solution?
  - [ ] `Kleisli / WriterT / WriterT / etc` => `A`
  - [x] `algWithGenericType`
  - [x] `algWithCurryMethod`
  - [x] Varargs
    - [x]  `AlgWithVarArgsParameter`
- **ApplyK** https://github.com/typelevel/cats-tagless/pull/516
  - [x] `A => F[_]` functions support 
  - [ ] Varargs
    - [ ]  `AlgWithVarArgsParameter`
- **InvariantK** https://github.com/typelevel/cats-tagless/pull/513
  - [x] `A => F[_]`
  - [ ] `A => U[F]`
  - [ ] `F[_] => F[_]`
  - [ ] `A => Kleisli / WriterT / WriterT / etc`
    - is there a generic solution?
  - [ ] `Kleisli / WriterT / WriterT / etc` => `A`
  - [ ] `NoEffectMethod`
  - [ ] `ContravariantEff`
  - [ ] `InvariantEff`
  - [ ] `AlgWithTypeMember`
  - [ ] Algebras with extra Type Parameters 
    - [ ] `WithExtraTpeParam`
    - [ ] `AlgWithExtraTP2` 
  - [ ] `AlgWithDef`
  - [ ] `AlgWithTypeParam`
  - [ ] `AlgWithCurryMethod`
  - [ ] Varargs
    - [ ]  `AlgWithVarArgsParameter`
  - [ ] `AlgWithByNameParameter`
- **ContravariantK** https://github.com/typelevel/cats-tagless/pull/518
  - [x] `A => F[_]`
  - [ ] `U[F]` type support in the functions result types 
    - (at this point only is supported `Cokleisli` only, is there a generic solution?
  - [ ] `TestAlgebraWithExtraTypeParam` 
  - [ ] ...
- **FunctionKLift** https://github.com/typelevel/cats-tagless/pull/425
  - [ ] Is it a cats citizen? An exposed `FunctionK.liftFunction` into Scala 3

This is a _very_ WIP PR to add Scala3 cross compilation. This PR covers a code structure change as well as the initial `FunctorK, SemigroupalK, ApplyK, InvariantK, ContravariantK` derivation macros implementations. 

The structure is changed to ensure that the Scala2 code compiles and runs as before; pay attention to `scala`, `scala-2`, and `scala-3` directories both in `src` and in `test`.

It can also be read as a port of everything I've played around with in terms of https://github.com/pomadchin/tagless-derivation, and reshaping it into a form of a PR.

Unfortunately a lot of test cases are not covered (i.e. curried functions, generic functions singatures, etc). There is still some significant work left to do to make at least these tests pass.

~~In addition the new `FunctionK` construct broke`cats.tagless.tests.autoSemigroupalKTests` and `cats.tagless.tests.autoContravariantKTests` a bit.~~ Fixed via b939f5d.

Feel free to jump in and push into this branch!

### CODE REVIEW

File: core/src/main/scala/cats/tagless/InvariantK.scala [hunk 1]
```diff
@@ -28,63 +29,67 @@ trait InvariantK[Alg[_[_]]] extends Serializable {
   def imapK[F[_], G[_]](af: Alg[F])(fk: F ~> G)(gk: G ~> F): Alg[G]
 }
 
-object InvariantK extends InvariantKInstances01 {
-  implicit def catsTaglessApplyKForEitherK[F[_], A]: ApplyK[EitherK[F, *[_], A]] =
-    eitherKInstance.asInstanceOf[ApplyK[EitherK[F, *[_], A]]]
+object InvariantK extends InvariantKInstances01 with DerivedInvariantK {
+  implicit def catsTaglessApplyKForEitherK[F[_], A]: ApplyK[({ type W[g[_]] = EitherK[F, g, A] })#W] =
+    eitherKInstance.asInstanceOf[ApplyK[({ type W[g[_]] = EitherK[F, g, A] })#W]]
```
Comment (pomadchin) [inline core/src/main/scala/cats/tagless/InvariantK.scala:34]: No fancy syntax to make both Scala 2 and Scala 3 compilers happy.

File: core/src/main/scala/cats/tagless/FunctionKLift.scala [hunk 2]
```diff
@@ -19,10 +19,11 @@ package cats.tagless
 import cats.arrow.FunctionK
 
 // https://github.com/typelevel/cats/issues/2553#issuecomment-493712879
+// https://github.com/typelevel/cats/blob/v2.9.0/core/src/main/scala-2/src/main/scala/cats/arrow/FunctionKMacros.scala
 object FunctionKLift {
```
Comment (pomadchin) [inline core/src/main/scala/cats/tagless/FunctionKLift.scala:23]: Mb it makes sense to expose FunctionK.liftFunction into Scala3 as well to facilitate backwards compat. 💭

**Comment by joroKr21** [comment]
I think it would be fine to not support all features in the beginning. At least the advantage of using macros is that we can add more features later without breaking bincompat.

> is there a generic solution?

Yes, see here: https://github.com/typelevel/cats-tagless/blob/master/macros/src/main/scala/cats/tagless/DeriveMacros.scala#L293-L300

```scala
          method.transformSubst(af, f -> g) {
            case Parameter(pn, pt, _) if occursIn(pt)(f) =>
              val F = method.summon[ContravariantK[Any]](polyType(f :: Nil, pt))
              q"$F.contramapK[$g, $f]($pn)($fk)"
          } { case delegate =>
            val F = method.summon[FunctorK[Any]](polyType(f :: Nil, method.returnType))
            q"$F.mapK[$f, $g]($delegate)($fk)"
          }
```

We are summoning the type class to do the heavy lifting and we have defined instances for monad transformers.

File: tests/src/test/scala-3/cats/tagless/tests/autoFunctorKTests.scala [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+
+  trait AlgWithNonEffectMethod[F[_]] derives FunctorK:
+    def a(i: Int): F[Int]
+    def b(i: Int): Int
+
+  // TODO: @finalAlg
+  // trait AlgWithTypeMember[F[_]] derives FunctorK {
+  //   type T
+  //   def a(i: Int): F[T]
+  // }
```
Comment (pomadchin) [inline tests/src/test/scala-3/cats/tagless/tests/autoFunctorKTests.scala:244]: Wondering, is it related to https://github.com/lampepfl/dotty/issues/15145 https://github.com/lampepfl/dotty/issues/8887, or it's a bad macro code.

```scala
class $anon() extends java.lang.Object with autoFunctorKTests.this.AlgWithTypeMember[G] {
  def list(id: scala.Int): G[autoFunctorKTests.this.AlgWithTypeMember[G].T] = fk.apply[autoFunctorKTests.this.AlgWithTypeMember[G].T](af.list(id))
}
```

**Comment by pomadchin** [comment]
@joroKr21 I can split this PR into several smaller onces if it makes sense, to unblock it 💭

**Comment by joroKr21** [comment]
@pomadchin if you can do that it would be amazing! 100 files to review is daunting. But I haven't had much time lately either. It's just been a mess for me with work but I've more or less sorted it out. I think in a few weeks I will be in a good position to do a more thorough review and get this in. Sorry for the delays and thank you for persisting.

**Comment by pomadchin** [comment]
@joroKr21 🥇 understood! haha out of time as well; I think I'll work on splitting into +- the following set of PRs:
- initial scala3 cross compilation (project setup, core cross compiled)
- separate macro PRs per type class; we shall see if we want a separate PR for a type class, but I can start with smth simple i.e. FunctorK and we'll figure out the strategy to approach it from there.

**Comment by kubukoz** [comment]
huge :+1: on splitting. `core` would be a really great start, `FunctorK` as the first typeclass to migrate seems awesome as well.

**Comment by pomadchin** [comment]
Hope next week I'll find some time :+1:

**Comment by joroKr21** [comment]
It's a daunting task to rebase this. Perhaps some of the early commits can be omitted now? Had for me to tell.

**Comment by pomadchin** [comment]
@joroKr21 no worries, I think i'll handle it; I have a plan: use the new main + cherry-pick the rest :D And then split it into small portions.

**Comment by joroKr21** [comment]
Let me know if I can help with something 👍

**Comment by pomadchin** [comment]
@joroKr21 I had to drop the whole PR history, but on the bright side it was an ez code backport. Commits history is preserved here: https://github.com/pomadchin/cats-tagless/tree/feature/scala-3-init-backup

I think that's the point I can use your help already!

I'll make the FunctorK macro PR soon (will appreciate if you'll pick it up after that), and this PR can be used as a code source for the future work.

**Comment by joroKr21** [comment]
> @joroKr21 I had to drop the whole PR history, but on the bright side it was an ez code backport. Commits history is preserved here: https://github.com/pomadchin/cats-tagless/tree/feature/scala-3-init-backup

Alright, sounds good 👍 

> I'll make the FunctorK macro PR soon (will appreciate if you'll pick it up after that), and this PR can be used as a code source for the future work.

Ok, I'll wait for that. Will be tomorrow though, it's getting late here 😄

**Comment by pomadchin** [comment]
I reimplemented SemigroupalK using the new awesome syntax; will make a PR today evening (EST) mb; + I'm thinking about including InvariantK into it.

**Comment by kubukoz** [comment]
> I reimplemented SemigroupalK using the new awesome syntax;

as a person who hasn't done that much macro stuff, which part is the new awesome syntax, @pomadchin?

**Comment by pomadchin** [comment]
@kubukoz check the [DerivedMacros](https://github.com/typelevel/cats-tagless/blob/0a295a0c4aa5c7e56141c7c0b636c93bd0633668/core/src/main/scala-3/cats/tagless/macros/DeriveMacros.scala#L27); some really marvellous stuff done by @joroKr21

**Comment by pomadchin** [comment]
ApplyK is next on the line, should be a low hanging fruit with what we have.

**Comment by pomadchin** [comment]
It took us more than a year to merge it in 😶‍🌫️ @joroKr21 rocks, the real macro stuff maestro!

Closing this one; it's been merged in terms of the following PRs:

* https://github.com/typelevel/cats-tagless/pull/425
* https://github.com/typelevel/cats-tagless/pull/442
* https://github.com/typelevel/cats-tagless/pull/513
* https://github.com/typelevel/cats-tagless/pull/516
* https://github.com/typelevel/cats-tagless/pull/518

**Comment by joroKr21** [comment]
Thank you for driving this @pomadchin! It's been a massive effort.

**Pull request closed without merging**